prune/prune.go

```diff
@@ -16,184 +16,209 @@ package prune
+func NewPruner(prunerClient client.Client, gvk schema.GroupVersionKind, strategy StrategyFunc, opts ...PrunerOption) (Pruner, error) {
+
+	if gvk.Empty() {
+		return Pruner{}, fmt.Errorf("error when creating a new Pruner: gvk parameter can not be empty")
 	}
 
-	if len(config.Namespaces) == 0 {
-		return fmt.Errorf("namespaces are required")
+	pruner := Pruner{
+		registry: defaultRegistry,
+		client:   prunerClient,
+		logger:   Logger(context.Background(), Pruner{}),
+		gvk:      gvk,
+		strategy: strategy,
 	}
 
-	if containsString(config.Namespaces, "") {
-		return fmt.Errorf("empty namespace value not supported")
+	for _, opt := range opts {
+		opt(&pruner)
 	}
 
-	_, err = labels.Parse(config.LabelSelector)
-	if err != nil {
-		return err
+	return pruner, nil
+}
+
+// Prune runs the pruner.
+func (p Pruner) Prune(ctx context.Context) ([]client.Object, error) {
+	var objs []client.Object
+	p.logger.Info("Starting the pruning process...")
+	listOpts := client.ListOptions{
+		LabelSelector: labels.Set(p.labels).AsSelector(),
+		Namespace:     p.namespace,
 	}
 
-	if config.Strategy.Mode == MaxAgeStrategy {
-		_, err = time.ParseDuration(config.Strategy.MaxAgeSetting)
+	var unstructuredObjs unstructured.UnstructuredList
+	unstructuredObjs.SetGroupVersionKind(p.gvk)
+	if err := p.client.List(ctx, &unstructuredObjs, &listOpts); err != nil {
+		return nil, fmt.Errorf("error getting a list of resources: %w", err)
+	}
+
+	for _, unsObj := range unstructuredObjs.Items {
+		obj, err := convert(p.client, p.gvk, &unsObj)
 		if err != nil {
-			return err
+			return nil, err
+		}
+
+		if err := p.registry.IsPrunable(obj, p.logger); isUnprunable(err) {
+			p.logger.Info(fmt.Errorf("object is unprunable: %w", err).Error())
+			continue
+		} else if err != nil {
+			return nil, err
 		}
+
+		objs = append(objs, obj)
 	}
-	if config.Strategy.Mode == MaxCountStrategy {
-		if config.Strategy.MaxCountSetting < 0 {
-			return fmt.Errorf("max count is required to be greater than or equal to 0")
+
+	objsToPrune, err := p.strategy(ctx, objs)
+	if err != nil {
+		return nil, fmt.Errorf("error determining prunable objects: %w", err)
+	}
+
+	// Prune the resources
+	for _, obj := range objsToPrune {
+		if err = p.client.Delete(ctx, obj); err != nil {
+			return nil, fmt.Errorf("error pruning object: %w", err)
 		}
 	}
-	return nil
+
+	return objsToPrune, nil
 }
 
 // Logger returns a logger from the context using logr method or Config.Log if none is found
 // controller-runtime automatically provides a logger in context.Context during Reconcile calls.
 // Note that there is no compile time check whether a logger can be retrieved by either way.
 // keysAndValues allow to add fields to the logs, cf logr documentation.
-func Logger(ctx context.Context, cfg Config, keysAndValues ...interface{}) logr.Logger {
+func Logger(ctx context.Context, pruner Pruner, keysAndValues ...interface{}) logr.Logger {
```

Comment:
I don't think this should be part of the exposed API. There should at most be a `pruner.Logger` function that returns the configured logger.